### BUSA3020 - Business Analytics - Assignment 1 {-}


---

**Assignment Points**: 100  
**Due Date**: Friday Week 4 (18 August) @ 11.59pm  
**Submission**: Submit your file using the submission link on iLearn


- Put **all your work** into this file;
- Failure to supply solutions in the cells provided below each question will result in a loss of 30 points;
- Follow all instructions closely and **do not** print your variables to the screen unless explicitly asked to do so;
    - Comment out print statements where necessary and make sure that your submitted notebook does not have the output of previously executed print statements;
    - 10 marks will be deducted for every redundant print statement not explicitly asked for.



### About the Assignment

Credit score cards are used as a risk control method in the financial industry. Personal information submitted by credit card applicants are used to predict the probability of future defaults. The bank employs such data to decide whether to issue a credit card to the applicant or not.




| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| Income   | Annual income  |  |
| Gender   | Applicant's Gender   | Male = 0, Female = 1  |
| Car | Car Ownership | Yes = 1, No = 0 | 
| Children | Number of Children | |
| Real Estate | Real Estate Ownership | Yes = 1, No = 0 
| Days Since Birth | No. of Days | Count backwards from current day (0), -1 means yesterday
| Days Employed | No. of Days | Count backwards from current day(0). If positive, it means the person is currently unemployed.
| Payment Default | Whether a client has overdue credit card payments | Yes = 1, No = 0



---


### Problem 1 - (50 points) {-}


**Question 1** 

- Import the `assignment_data.xlsx` file from `data` folder into a pandas DataFrame named `df`; 
- Delete duplicate rows from `df` according to `ID`;
- Delete the `ID` column.
- How many rows are left in `df`?

(10 points)

In [28]:
import pandas as pd

import matplotlib.pyplot as plt

import numpy as np


In [29]:
column_names=['ID','Gender','Car','Real Estate','Children','Income','Days Since Birth','Days Employed','Payment Default']
df=pd.read_excel("assignment_data.xlsx", names = column_names)

In [30]:
df.drop_duplicates(subset='ID', keep= 'first', inplace= True)

Dataframe originally has 10000 rows, after deleting duplicate rows, there are 5976 rows. 


---

**Question 2**

- Reset the index in `df` using an appropriate function from `pandas` so that the new index corresponds to the number of rows (make sure to delete the old index). 
- How many positive values of `Days Employed` are there?
- Replace the positive values of `Days Employed` with 0 (zero) in `df`

(10 points)

In [8]:
df.reset_index() #### ASK HELP

index       ID  Gender  Car  Real Estate  Children    Income  \
0         0  5100111       0    1            0         0  135000.0   
1         1  5091281       0    0            1         0  202500.0   
2         2  5116669       1    0            1         0  157500.0   
3         3  5143035       0    0            0         1  270000.0   
4         4  5142498       1    1            1         1  135000.0   
...     ...      ...     ...  ...          ...       ...       ...   
5971   9992  5038548       0    1            0         0  225000.0   
5972   9993  5116921       0    1            1         2  211500.0   
5973   9994  5096414       1    0            0         0   76500.0   
5974   9995  5116605       0    1            1         0  180000.0   
5975   9999  5061559       1    0            1         0  202500.0   

      Days Since Birth  Days Employed  Payment Default  
0               -10239         365243                1  
1               -13645          -5374                1  
2               -16628          -3654                1  
3               -10108          -1473                1  
4               -15481          -2874                0  
...                ...            ...              ...  
5971             -9479          -1143                1  
5972            -10435          -2987                1  
5973            -16876            -97                0  
5974            -17115          -4599                0  
5975            -15047          -1720                1  

[5976 rows x 10 columns]

In [9]:
positive_number= np.sum(df['Days Employed']>0)#### double check
print(positive_number)

967


There are 967 positive values in Days Employed in df. 

In [10]:
df.loc[df['Days Employed'] > 0, 'Days Employed'] = 0

---
**Question 3**

Create two new variables in `df` named 

1. `Age`;
2. `Years in Employment`,

which measure age and employment length in **years** (decimal numbers) from `Days Since Birth` and `Days Employed` by applying approapriate transformations on these variables. 

Delete the original variables `Days Since Birth` and `Days Employed`.

(5 points)


In [11]:
df['Age']= abs(df['Days Since Birth']/365)
df['Years in Employment']= abs(df['Days Employed']/365)

In [12]:
df= df.drop('Days Since Birth',axis=1)
df=df.drop('Days Employed', axis =1)

---
**Question 4**

- Create a **one**-dimensional NumPy array named `y` by exporting the first 5,000 observations of `Payment_Default`. (Hint: see `ravel()` function)
- Create a NumPy array named `X` by exporting the first 5,000 observations of the following columns `Gender`, `Car`, `Real Estate`, `Children`, `Income`, `Age`, `Years in Employment`.
 
(10 points)


In [13]:
y=df['Payment Default'].head(5000).to_numpy().ravel()

In [14]:
Columns=['Gender','Car','Real Estate','Children','Income','Age','Years in Employment']
X=df[Columns].head(5000).to_numpy()

---

**Question 5** 

- Use an appropriate `scikit-learn` library we learned in class to create the following NumPy arrays: `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% training and 30% test datasets. 
- Set `random_state` to 0 and stratify subsamples so that train and test datasets have roughly equal proportions of the target's class labels. 

(5 points) 

In [15]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 1, stratify = y)


In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0, stratify = y)

---

**Question 6**

- Create new variables by using an appropriate `scikit-learn` library we learned in class to standardize the features from the training and test datasets to mean zero and variance one. Name the new variables by appending '_scaled' to the original variable names.


(10 points)   

In [17]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()

X_train_scaled = sc.fit_transform(X_train)


X_test_scaled = sc.transform(X_test)

Column_name = ['Gender', 'Car', 'Real Estate', 'Children', 'Income', 'Age', 'Years in Employment']
Column_name_scaled = [name + '_scaled' for name in Column_name]

X_train_scaled_df = pd.DataFrame(X_train_scaled, columns=Column_name_scaled)
X_test_scaled_df = pd.DataFrame(X_test_scaled, columns=Column_name_scaled)



---

## Problem 2 - (20 Points) {-}

**Question 7**

Fit the following two classifiers to the transformed training dataset using `scikit-learn` libraries.

- Perceptron - name your instance `pc` set `random_state=1`
- Logistic Regression - name your instance `lr` set `random_state=1`

When initializing instances of the above classifiers only set the parameters referenced above and nothing else.

(20 points)

In [18]:
from sklearn.linear_model import Perceptron
pc=Perceptron(random_state=1)
pc.fit(X_train_scaled_df, y_train)


from sklearn.linear_model import LogisticRegression
lr=LogisticRegression(random_state=1)
lr.fit(X_train_scaled_df, y_train)


LogisticRegression(random_state=1)

---
## Problem 3 - (30 points) {-}


**Question 8**

- Using a method built into each of the two classifiers compute their prediction accuracies on the training data;
- Store the accuracy values into variables named according to the following pattern: `classifier_name_accuracy_train`, e.g. you should have `lr_accuracy_train`; 
- Print the two accuracy **variables** along with their brief descriptions.

(10 points)

In [19]:
from sklearn.metrics import accuracy_score

pc_predictions_train = pc.predict(X_train_scaled_df)


In [20]:
pc_accuracy_train = accuracy_score(y_train, pc_predictions_train)

In [21]:
lr_predictions_train = lr.predict(X_train_scaled_df)

In [22]:
lr_accuracy_train = accuracy_score(y_train, lr_predictions_train)

In [23]:
print("Perceptron Accuracy on Training Data:", pc_accuracy_train)
print("Logistic Regression Accuracy on Training Data:", lr_accuracy_train)

Perceptron Accuracy on Training Data: 0.496
Logistic Regression Accuracy on Training Data: 0.5577142857142857


---

**Question 9** 

- Using a method built into each of the above classifiers compute their prediction accuracy for the test dataset
- Store the accuracy values into variables named according to the following pattern: `classifier_name_accuracy_test`, e.g. you should have `lr_accuracy_test`. 
- Print the two accuracy **variables** along with brief descriptions.

(10 points)

In [24]:
pc_predictions_test = pc.predict(X_test_scaled_df)
pc_accuracy_test = accuracy_score(y_test, pc_predictions_test)

In [25]:
lr_predictions_test = lr.predict(X_test_scaled_df)
lr_accuracy_test = accuracy_score(y_test, lr_predictions_test)

In [26]:
print("Perceptron Accuracy on Test Data:", pc_accuracy_test)
print("Logistic Regression Accuracy on Test Data:", lr_accuracy_test)

Perceptron Accuracy on Test Data: 0.49466666666666664
Logistic Regression Accuracy on Test Data: 0.5073333333333333


---

**Question 10** 

Using nicely formated text in Markdown comment on the accuracies computed in Questions 8 & 9 making sure you address:
- training and test set datasets; 
- Perceptrion and Logistic Regression models. 

Are the results as expected, and why or why not? (Hint: You are not expected to comment on why a particular model is better.) 

(10 marks)


|  Models            |  Training| Test |
| :---------------- | :------: | ----: |
| Perceptrion        |   0.496   | 0.558 |
| Logistic Regression model |   0.494   | 0.507 |


### Accuracy for the Perceptron Model:

#### Training Set Accuracy:
The Perceptron model had an accuracy of 0.496 in the training dataset, demonstrating how well the model performed on the trained data.

#### Test Set Accuracy:
On the testing dataset, the model had an accuracy of 0.558. 

These accuracy scores for this Perceptron model are as expected. This data is a dataset that works well with the simple linear classifier like Perceptron model.

### Accuracy for the Logistic Regression Model:

#### Training Set Accuracy:
For the Logistic Regression model, the training set accuracy was 0.494, showcasing how well the model fits the training data while taking into account the probabilities of different classes.

#### Test Set Accuracy:
The test set accuracy of the Logistic Regression model was 0.507. 

The accuracy results for the Logistic Regression model are also consistent with our expectations. This is because, the features and the target variable in this dataset works well with this binary classification, the is the Logistic Regression model.

In conclusion, the accuracies that was resulted for both the Perceptron and Logistic Regression models are as expected. 
based on the nature of the models and the characteristics of the datasets.